### PR TITLE
Initial sequencing spec

### LINKED ISSUE #183: Spec sequencer window

### PR DESCRIPTION
**Pull request by protolambda**
This PR:
- Updates the rollup node doc:
  - define seq window (Fixes #183)
  - define seq batches
  - update sync (same, but need full seq window of blocks after finding common point)
  - misc updates, no more single block output, no deposit-blocks, etc.
- Updates the glossary with sequencing terms, and remove outdated "deposit block"

This PR does not spec the job of the sequencer to *produce* the data, this should probably go into a separate document, similar to how `proposing.md` is split out.

**Metadata**
- Fixes #183 
- Fixes ENG-1978

### CODE REVIEW

File: specs/rollup-node.md [hunk 1]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+input. The window is identified by an `epoch`, equal to the block number of the first block in the window.
+
+As the full derivation of the L2 chain by the driver progresses each derivation step shifts the window forward by a
+single L1 block: the windows overlap.
+
+Each sequencing window is derived into a variable number of L2 blocks, depending on the timestamps of L1 and L2.
+
+The L2 has a fixed block time and no more than one batch per block,
+meaning that gaps between the batches (ordered by timestamp) are interpreted as batches with empty transaction-lists,
+thus construing L2 blocks that only contain deposit transaction(s).
```
Comment (tynes) [inline specs/rollup-node.md:91]: is this supposed to be `constructing`?

File: specs/rollup-node.md [hunk 2]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+    - Batches not matching filter criteria are ignored:
+      - `batch.epoch == sequencing_window.epoch`, i.e. for this sequencing window
+      - `(batch.timestamp - genesis_l2_timestamp) % block_time == 0`, i.e. timestamp is aligned
+      - `min_l2_timestamp < batch.timestamp < max_l2_timestamp`, i.e. timestamp is within range
+      - The batch is the first batch with `batch.timestamp` in this sequencing window,
+        i.e. one batch per L2 block number
 
 [random]: https://eips.ethereum.org/EIPS/eip-4399
 
+Batches are formatted as: `type ++ RLP([epoch, timestamp, transaction_list])`
```
Comment (tynes) [inline specs/rollup-node.md:134]: When being abstract, should it be `type ++ blob` where type 0 is `type ++ RLP([epoch, timestamp, transaction_list])` ?

**Review by trianglesphere — COMMENTED** [review]
One of the things that I read in the full specs (not changed here) is that we do not look at the transactions nonce.

On that vein, one of the things that I don't see addressed is how to deal with invalid transactions. Transactions tend to depend on previous transactions, so what happens when a batch submission is missed (all sorts of reasons for it to not be submitted or the chain could re-org). That's a big area that I would like to see addressed more.

I've also left some comments on the timestamps. I'm probably more cautious of them than most, but I think that we could have them with some other sanity checks as well.

File: specs/rollup-node.md [hunk 3]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+
+- `type` is a single `byte` identifying the batch format
+- `epoch` is the sequencing window epoch, i.e. the first L1 block number
+- `timestamp` is the L2 timestamp of the block
+- `transaction_list` is an RLP encoded list of [EIP-2718] encoded transactions
```
[the remark this answers is not in the document]
Comment (tynes) [reply specs/rollup-node.md:139]: I don't think we should block legacy transactions - some existing tooling might break

[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:139]: All transaction types (well, except deposits) are valid here. EIP-2718 also covers legacy transactions.

File: specs/rollup-node.md [hunk 4]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+  - derived from [receipts][g-receipts]: zero or more *[user-deposited transactions][g-user-deposited]*.
+- *[sequenced transactions][g-sequencing]*: derived from [sequencer batches][g-sequencer-batch],
+  zero or more regular transactions, signed by L2 users.
 
 ------------------------------------------------------------------------------------------------------------------------
 
+## From L1 chain to Sequencing Window
+
+A [sequencing window][g-sequencing-window] is a fixed number consecutive L1 blocks that a derivation step takes as
+input. The window is identified by an `epoch`, equal to the block number of the first block in the window.
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:82]: Yes, that is correct

File: specs/rollup-node.md [hunk 5]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+As the full derivation of the L2 chain by the driver progresses each derivation step shifts the window forward by a
+single L1 block: the windows overlap.
+
+Each sequencing window is derived into a variable number of L2 blocks, depending on the timestamps of L1 and L2.
+
+The L2 has a fixed block time and no more than one batch per block,
+meaning that gaps between the batches (ordered by timestamp) are interpreted as batches with empty transaction-lists,
+thus construing L2 blocks that only contain deposit transaction(s).
+
+The L2 blocks produced by a sequencing window are bounded by timestamp:
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:93]: - There is indeed a minimum requirement, it's relatively implicit now though. The bound is at least one L2 block-time large. This is where deposits will get processed.
- The L1 bounds limit L2 sufficiently already. We could put a hardcap of `seq_window_time / l2_time + margin` blocks of output, but that has not been discussed or implemented yet. Better for a follow-up PR IMHO.

File: specs/rollup-node.md [hunk 6]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+
+- `min_l2_timestamp = prev_l2_timestamp + l2_block_time`
+- `max_l2_timestamp = l1_timestamp + l2_block_time`, where `l1_timestamp` is the timestamp of the
+  first L1 block of the sequencing window. (maximum bound, may not be aligned with block time)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:97]: Hmm, I may have messed up these timestamp bounds, will search back and check.

[the remark this answers is not in the document]
Comment (trianglesphere) [reply specs/rollup-node.md:97]: So this makes sense from the implementation POV, but does have some weird behavior.

File: specs/rollup-node.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (trianglesphere) [reply specs/rollup-node.md:139]: Yep, turns out it does. May want to explicitly disallow deposits b/c 2718 doesn't reference which typed transactions are allowed.

File: specs/rollup-node.md [hunk 7]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+
+If there are no batches present in the sequencing window then the L2 chain is extended up to `max_l2_timestamp` (incl.)
+with empty batches, but otherwise regular block derivation.
+
+Note that with short block times on L1 the L2 time may increment beyond the L1 time,
+but the longer target block time of L1 will correct back and allow the timestamps to align again.
+
```
Comment (norswap) [inline specs/rollup-node.md:104]: Imho this section is the weak part of the spec, I think we can better explain the relation between L1 and L2 blocks.

No particular suggestion though (any such suggestion would probably be equivalent to me trying to write an alternative explanation, which is probably better done in another PR.)

**Review by norswap — COMMENTED** [review]
Review focused on the diff.

Might do my own pass later (in a new PR) for style and reworking some explanations, but this is pretty good imho 👍

File: specs/rollup-node.md (hunk 1 above)
Comment (protolambda) [reply specs/rollup-node.md:91]: That would also work, but it's used more like "result of interpretation", which "construe" captures too.

File: specs/rollup-node.md (hunk 2 above)
Comment (protolambda) [reply specs/rollup-node.md]: Updated this to word it better and also include encoding of multiple batches (i.e. when we are submitting multiple L2 blocks together)

File: specs/rollup-node.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md]: [ac1708b](https://github.com/ethereum-optimism/optimistic-specs/pull/223/commits/ac1708b830c0cd1fac686d9d9bc6c66f1302bddc) filters out batches that include deposit-type transactions.

File: specs/glossary.md [hunk 8]
```diff
@@ -140,6 +143,38 @@ on-chain-confirmed head, or the on-chain-finalized head.
...
+
+A sequencing epoch is a number identifying the start of a [sequencing window](#sequencing-window),
+equal to the L1 block number of the first block in the window.
+
+## Sequencer batch
+
+[sequencer-batch]: glossary.md#sequencer-batch
+
+A sequencer batch is list of L2 transactions tagged with an [`epoch`](#sequencing-epoch) and L2 block `timestamp`.
+Each L2 block can have one batch of transactions, an input for the block derivation.
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/glossary.md:174]: Added a description of encoding/decoding of bundles of batches. When deriving from a L1 transaction, multiple batches can be read now, and the same filtering etc. still applies to them.

The job of submitting bundled batches should go into a separate doc, `sequencing.md` or something, similar to the separate `proposals.md`. There we can call out the reverse process, how multiple L2 blocks land into the same L1 transaction data.

File: specs/rollup-node.md [hunk 9]
```diff
@@ -59,31 +64,82 @@ currently only concerned with the specification of the rollup driver.
...
+- Of the *first* block in the window only:
+  - L1 block attributes:
+    - block number
+    - timestamp
+    - basefee
+    - *random* (the output of the [`RANDOM` opcode][random])
+  - L1 log entries emitted for [user deposits][g-deposits], augmented with
+    - `blockHeight`: the block-height of the L1 block
+    - `transactionIndex`: the transaction-index within the L2 transactions list
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:119]: This also relates to the first point you are making, I'll clarify that deposits are put at the start of the transactions list.

File: specs/rollup-node.md [hunk 10]
```diff
@@ -96,6 +152,11 @@ entries.
 
 [deposit-contract-spec]: deposits.md#deposit-contract
 
+Each of the derived `PayloadAttributes` starts with a L1 Attributes transaction.
+
+The [User-deposited] transactions are all put in the first of the derived `PayloadAttributes`,
+inserted after the L1 Attributes transaction, before any [sequenced][g-sequencing] transactions.
+
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:173]: The reason is to simplify L2 sync, by making it always possible to read the L1 information from the block-data, instead of reading from some L2 state. The nice thing is that we do not have to batch-submit this part, it's just part of the derivation process, so it only contains us a little bit of L2 disk storage.

File: specs/deposits.md [hunk 11]
```diff
@@ -76,7 +75,7 @@ for a variable-length encoding scheme.
 ### Kinds of Deposited Transactions
 
 Although we define only one new transaction type, we can distinguish between two kinds of deposited
-transactions, based on their positioning in the [deposit block][g-deposit-block]:
+transactions, based on their positioning in the L2 block:
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/deposits.md:78]: fixed in [f8ccf4a](https://github.com/ethereum-optimism/optimistic-specs/pull/223/commits/f8ccf4a14b89029deefa20a3c004185de56daff4), feel free to change this in a next PR

File: specs/rollup-node.md [hunk 12]
```diff
@@ -41,6 +45,7 @@ currently only concerned with the specification of the rollup driver.
 **Table of Contents**
 
 - [L2 Chain Derivation](#l2-chain-derivation)
+  - [From L1 chain to Sequencing Window](#from-l1-chain-to-sequencing-window)
   - [From L1 Blocks to Payload Attributes](#from-l1-blocks-to-payload-attributes)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:49]: I like to separate the processes of selecting the window and deriving payload attributes. You can try a different structure in a next PR here if you like

File: specs/rollup-node.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:119]: Clarified the deposits-at-start point elsewhere now, and think "transaction-index within the L2 transactions list" here is sufficient

File: specs/rollup-node.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/rollup-node.md:175]: Added rationale in [f8ccf4a](https://github.com/ethereum-optimism/optimistic-specs/pull/223/commits/f8ccf4a14b89029deefa20a3c004185de56daff4)

**Review by mslipper — APPROVED** [review]

**Comment by mslipper** [comment]
LGTM. I think we can discuss the rest async and PR those changes into additional PRs.
